src/ParallelAlgorithms/Interpolation/Callbacks/ObserveSurfaceData.hpp

```diff
@@ -32,10 +35,100 @@
+/// columns for each row.
+///
+/// \param legend the legend to fill for the strahlkorper data we want to write
+/// \param data the spherical harmonic data to fill that we want to write
+/// \param strahlkorper the strahlkorper from which to get the spherical
+/// harmonic data to fill `legend` and `data`
+/// \param time the current time
+/// \param max_l the maximum value that \f$l_max\f$ could be (sets number of
+/// columns to write)
+template <typename Frame>
+void fill_ylm_legend_and_data(
+    const gsl::not_null<std::vector<std::string>*> legend,
+    const gsl::not_null<std::vector<double>*> data,
+    const ylm::Strahlkorper<Frame>& strahlkorper, const double time,
+    const size_t max_l) {
+  const size_t l_max = strahlkorper.l_max();
+  ASSERT(max_l >= l_max,
+         "The Lmax of the Ylm data to write ("
+             << l_max << ") is larger than the maximum value that l can be ("
+             << max_l << ").");
+  const std::array<double, 3> expansion_center =
+      strahlkorper.expansion_center();
+  // number of terms in
+  // \sum_{l=0}^{max\_l} \sum_{m=-l}^{l} F^{lm} Y^{lm}(\theta,\phi) is the
+  // sum of the first (max_l + 1) odd numbers, which is (max_l + 1)^2
+  const size_t num_coefficients = square(max_l + 1);
+  // time + 3 dims of expansion center + Lmax = 5 columns
+  const size_t num_columns = num_coefficients + 5;
+
+  legend->reserve(num_columns);
+  data->reserve(num_columns);
+
+  legend->emplace_back("Time");
+  data->emplace_back(time);
+  legend->emplace_back("ExpansionCenter_x");
+  data->emplace_back(expansion_center[0]);
+  legend->emplace_back("ExpansionCenter_y");
+  data->emplace_back(expansion_center[1]);
+  legend->emplace_back("ExpansionCenter_z");
+  data->emplace_back(expansion_center[2]);
+  legend->emplace_back("Lmax");
+  data->emplace_back(l_max);
+
+  const DataVector ylm_coefficients = strahlkorper.coefficients();
+  // fill coefficients for l in [0, l_max]
+  // l_max == m_max
+  ylm::SpherepackIterator iter(l_max, l_max);
+  for (size_t l = 0; l <= l_max; l++) {
+    for (int m = -l; m < 0; m++) {
+      legend->push_back(MakeString{} << "Im(" << l << "," << m << ")");
+
+      iter.set(l, m);
+      data->push_back(ylm_coefficients[iter()]);
+    }
+    for (int m = 0; m <= static_cast<int>(l); m++) {
+      legend->push_back(MakeString{} << "Re(" << l << "," << m << ")");
+
+      iter.set(l, m);
+      data->push_back(ylm_coefficients[iter()]);
+    }
+  }
+  // fill coefficients for l in [l_max + 1, max_l],
+  // i.e. higher order coefficients beyond this Strahlkorper's l_max == 0.0
+  data->resize(num_columns, 0.0);
+  for (size_t l = l_max + 1; l <= max_l; l++) {
+    for (int m = -l; m < 0; m++) {
+      legend->push_back(MakeString{} << "Im(" << l << "," << m << ")");
+    }
+    for (int m = 0; m <= static_cast<int>(l); m++) {
+      legend->push_back(MakeString{} << "Re(" << l << "," << m << ")");
+    }
+  }
+
+  ASSERT(legend->size() == data->size(),
```

Comment:
Move this check to the unit tests.

Reply:
Just removed this because it is already `CHECK`ed in the unit test